### PR TITLE
Generate access token on start up, return authenticated `httpClient`.

### LINKED ISSUE #6: Add in `bedrock.ready` to module configuration.
> I added in the flag because we only want the behavior (unlimited retries and exiting if we can't get the token) on start up...

Right, but that's a use case, not a description of the feature. There may be other use cases in the future. If we're convinced that that is unlikely, we should consider if other designs would be better.

To that end, it seems that what we're really trying to do here is avoid duplicating error handling code in numerous places by pushing it into this library. However, we're still adding event handling code everywhere to call into this library. Therefore, I'm still thinking we want this library to handle obtaining its initial access token on startup on its own -- and exit on its own -- removing the need for the flag entirely.

Why don't we put the `bedrock.ready` handler code into this module and allow this module to be configured with N many target servers (for which it will obtain N many access tokens on startup)?

_Originally posted by @dlongley in https://github.com/digitalbazaar/bedrock-oauth2-client/pull/5#discussion_r707502154_

### PR DESCRIPTION
**Pull request by mandyvenables**

### CODE REVIEW

File: CHANGELOG.md
Comment (dlongley) [inline CHANGELOG.md]:
```suggestion
- Added `createAuthzHttpClient` which returns an instance of `httpClient` with the  
```

File: lib/authHttpClient.js
Comment (dlongley) [inline lib/authHttpClient.js]: We're creating an instance here and it seems like this is an authorization (vs. authentication) http client so let's throw in the `z` so that's a bit more clear:

```suggestion
export function createAuthzHttpClient({oAuth2Client} = {}) {
```

File: lib/authHttpClient.js
Comment (dlongley) [inline lib/authHttpClient.js]:
```suggestion
  // create the authz header using the token
```

File: lib/authHttpClient.js
Comment (dlongley) [inline lib/authHttpClient.js]:
```suggestion
  // add the authz header to the headers object of args
```

File: lib/authHttpClient.js
Comment (dlongley) [inline lib/authHttpClient.js]: What if an `Authorization` header is already set? Can this check that and convert it to an array and append to it in that case?

File: lib/authHttpClient.js
Comment (dlongley) [inline lib/authHttpClient.js]: I imagine the number of retries may change over time or become configurable -- and I don't think it needs to be surfaced in the error message. So an error message of "Could not get access token." would be better. Also, I don't think the `BedrockError` constructor takes an `error` (presumably meant to be the "cause" here) as the second parameter, so that needs to be fixed.

```suggestion
  throw new BedrockError('Could not get access token.', error);
```

File: lib/main.js
Comment (dlongley) [inline lib/main.js]: It's generally a good idea to differentiate between authn and authz.

```suggestion
import {getAuthzClients} from './oauth2-client.js';
```

File: lib/oauth2-client.js
Comment (dlongley) [inline lib/oauth2-client.js]: Hmm, I'm not sure what I think of this usage. If two different calls are made with the same client ID but different other params ... then one of them may resolve to a result that used different params. Is that a problem?

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/oauth2-client.js
Comment (mandyvenables) [reply lib/oauth2-client.js]: @dmitrizagidulin Thoughts?

File: lib/authHttpClient.js
Comment (mandyvenables) [reply lib/authHttpClient.js]: Updated.

File: lib/authHttpClient.js
Comment (mandyvenables) [reply lib/authHttpClient.js]: Updated, is there a recommended error this should be? I set it as a `DataError` because I didn't think a `NotFoundError` was appropriate in this case.

File: lib/authHttpClient.js
Comment (dlongley) [inline lib/authHttpClient.js]: Actually, I don't think this code path is ever hit. If `getNewAccessToken` fails after `maxRetries`, it throws. There's no case where it returns an empty/invalid `accessToken`, right? If so, we should just let it throw and we don't actually care about the return value.

However, looking at this, it seems like it would retry POST operations (not just GET / idempotent operations). That seems like it could be a bad idea. Were we previously automatically retrying POST operations when a token would fail?

File: lib/main.js
Comment (dlongley) [inline lib/main.js]: This should be done in parallel to speed up startup time.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/authHttpClient.js
Comment (dlongley) [reply lib/authHttpClient.js]: The code path actually isn't hit (I think), so we can just remove it and simplify a bit above.

File: lib/oauth2-client.js
Comment (mandyvenables) [reply lib/oauth2-client.js]: @dlongley would creating a new key which is a string of `client_id` + `token_endpoint` be a good solution?

File: lib/authHttpClient.js
Comment (mandyvenables) [reply lib/authHttpClient.js]: Correct, this code path is not used. Removed.

File: lib/oauth2-client.js
Comment (dlongley) [reply lib/oauth2-client.js]: @mandyvenables the key should be whatever uniquely identifies a particular call to get the access token. If the caller ever receives an access token that doesn't match the parameters they passed in an unexpected/acceptable way, then there's a problem.

I think, at a minimum, the `token_endpoint` needs to be included because we mentioned that a client ID is only unique with respect to a particular endpoint. @dmitrizagidulin -- what other characteristics are there here for uniqueness and expectation?

File: lib/authHttpClient.js
Comment (mandyvenables) [reply lib/authHttpClient.js]: @dlongley The immediate intended use of the library is to re-try POST operations, and that was previously occurring.

Comment (dlongley) [reply lib/authHttpClient.js]: Ok, since this has been previously vetted and that's the main purpose, I'll resolve this particular thread.

File: lib/authHttpClient.js
Comment (dlongley) [inline lib/authHttpClient.js]: I recommend the file name be authzHttpClient too to match.

File: lib/oauth2-client.js
Comment (dmitrizagidulin) [reply lib/oauth2-client.js]: So, this is tricky. The `client_id` typically has a 1-1 mapping to the `token_endpoint` (meaning, if you need to go to a different `token_endpoint`, it generally means that it's a different zone of trust, and you'll be issued a different `client_id`). 

(This is, at least, for the typical OAuth2 setup, where mobile clients & third-party apps register with the authorization server, and receive their individual `client_id` and `client_secret` strings. It doesn't really hold for globally unique client_ids (like, WebIDs or DIDs) -- it's just not a use case the community has dealt with much.)

It's not inconceivable -- the protocol does not forbid reusing the client_id on multiple token endpoints.
So, I could see it either way (either just memoizing solely by `client_id`, or the `${client_id}${token_endpoint}` tuple.

Comment (dmitrizagidulin) [reply lib/oauth2-client.js]: Yeah, the more I think about it -- let's just memoize by client_id + endpoint.

Comment (dlongley) [reply lib/oauth2-client.js]: If it's possible that `client_id` may be the same across `token_endpoint`s then it's imperative that it be included in the key, otherwise it's a security risk. Additionally, the key must have a delimiter that isn't present in either of those values (presuming that they are the only values that need to go into the key, I'm still not sure). So, whatever values need to go into the key should probably just be bundled up using `JSON.stringify([a, b, c...])` unless we've got a better idea for delimiting.

Back to uniqueness -- do any of the other parameters need to be considered as part of the key?

Comment (mandyvenables) [reply lib/oauth2-client.js]: @dlongley I think the combination of `client_id` and `token_endpoint` should be sufficient, I can't think of any use cases where there would be identical `client_id`s  with different secrets on the same endpoint.

Comment (dlongley) [reply lib/oauth2-client.js]: @mandyvenables,

We have these params to deal with:

> client_id, client_secret, token_endpoint, grant_type, scope, maxRetries

So these are some of the questions I'd ask:

If I call the function with this:

`client_id='a', client_secret='b', token_endpoint='c', grant_type='g1', scope='s1'`

And then I later call the function with this:

`client_id='a', client_secret='b', token_endpoint='c', grant_type='differentGrantType', scope='s1'`

I will get back the same token from the first call. Is that ok?

What if I later call this:

`client_id='a', client_secret='b', token_endpoint='c', grant_type='g1', scope='differentScope'`

I will get back the same token from the first call. Is that ok?

What if I later call this:

`client_id='a', client_secret='b', token_endpoint='c', grant_type='differentGrantType', scope='differentScope'`

I will get back the same token from the first call. Is that ok?

What if I later call this:

`client_id='a', client_secret='differentSecret', token_endpoint='c', grant_type='whatever1', scope='whatever2'`

I will get back the same token from the first call. Is that ok?

Are all of these ok? Essentially, permute all of the options where only the `client_id` and `token_endpoint` don't change. Does it ever create a problem? If not, we're good. Otherwise, we've got a (potentially serious) problem.

Comment (mandyvenables) [reply lib/oauth2-client.js]: @dlongley based on our original use case, I don't see the scope or additional parameters changing, however if they did, the best course of action would be to generate a new access token. I have updated the key to include `grant_type` and `scope`. @dmitrizagidulin

Comment (dlongley) [reply lib/oauth2-client.js]: @mandyvenables, yeah, one current use case may not change these params, but we are designing a low level API here so we have to make sure that any valid use of that API won't create a problem for the user (who may have other use cases).

File: lib/oauth2-client.js
Comment (dlongley) [inline lib/oauth2-client.js:59]:
```suggestion
  // it is ok to return the same cached token when the params `client_secret` or
  // `maxRetries` are different, but these params below must be unique per token
  const key = JSON.stringify([client_id, token_endpoint, grant_type, scope]);
```

Comment (mandyvenables) [reply lib/oauth2-client.js:59]: Added.

File: lib/oauth2-client.js
Comment (mattcollier) [inline lib/oauth2-client.js:59]: I think a comment here indicating an array is used here (instead of an object) to ensure that `JSON.stringify` preserves the order.  But it does beg the question, why not a string?  According to the jsdocs, all of these elements are strings themselves.
```js
const key = `${client_id}${token_endpoint}${grant_type}${scope}`;
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#issue_with_plain_json.stringify_for_use_as_javascript

Comment (mandyvenables) [reply lib/oauth2-client.js:59]: @dlongley requested the `JSON.stringify` to delineate the different parameters.

File: lib/oauth2-client.js
Comment (dmitrizagidulin) [reply lib/oauth2-client.js]: Ok, so, let's examine all those parameters.
I think we agree at this point that `client_id` + `token_endpoint` is our initial unique caching key. Let's look at the others.

`client_secret` -- not useful. It's not allowed to have multiple registrations for the same `client_id` but with different `client_secret`. In the _rare_ occurrence that a client re-registers with the same client id (or updates their secret), then the RS (resource server) will either accept the old token (in which case, the cache will work without `client_secret`), or it won't (in which case, the cache doesn't work, and the token gets re-requested -- a price of another http roundtrip, but again, this is a rare occurrence that we don't need to optimize for, cache-wise).

`grant_type` -- not useful. Grant type has to do with how the client authenticated / requested the token in the first place. But by the time the logic gets down to issuing the token, the grant type is forgotten / does not matter. Put a different way, a client could reasonably get the same token regardless of which grant_type it used to request it.

`scope` -- This one's the closest candidate, as far as our composite caching key. Again, though, also not that useful. (Our client management libraries (or anybody's, really), don't juggle different tokens (from the same issuer) based on scope.

File: lib/authzHttpClient.js
Comment (davidlehn) [inline lib/authzHttpClient.js]: Looks like this flow can be simplified.  Also skipping the `await` as in `_handleTokenError`.
```suggestion
  try {
    return target.apply(thisArg, [url, options]);
  } catch(error) {
    return _handleTokenError({
      error, target, thisArg, url, options, oAuth2Client
    });
  }
```

Comment (mandyvenables) [reply lib/authzHttpClient.js]: Thanks, updated!

File: lib/oauth2-client.js
Comment (dlongley) [reply lib/oauth2-client.js:59]: If you use a string then you need delimiters that don't appear in the strings, otherwise evil like this can happen:

```
client_id1 = foo
token_endpoint1 = abc
...
token1 = get(client_id1, token_endpoint1, ...)

...

client_id2 = fooab
token_endpoint2 = c
...
token2 = get(client_id2, token_endpoint2, ...)

// oh no!
token1 === token2
```

:sob: 

It's similar to an SQL-injection attack.

File: lib/authzHttpClient.js
Comment (dlongley) [inline lib/authzHttpClient.js]: The header may already be an array:

```suggestion
    if(!Array.isArray(options.headers.Authorization)) {
      options.headers.Authorization = [options.headers.Authorization];
    }
```

File: lib/authzHttpClient.js
Comment (dlongley) [inline lib/authzHttpClient.js]: If `target.apply` returns a promise then it won't settle within the `try/catch` so `_handleTokenError` won't happen. We need to await it:

```suggestion
    return await target.apply(thisArg, [url, options]);
```

I think if we don't have a test catching this problem we should add one.

File: lib/authzHttpClient.js
Comment (dlongley) [inline lib/authzHttpClient.js]: `getNewAccessToken` always returns an access token or throws, right? If so:

```suggestion
  await getNewAccessToken({...oAuth2Client, maxRetries: 3});
  return target.apply(thisArg, [url, options]);
```

File: lib/storage.js
Comment (dlongley) [inline lib/storage.js:41]: It doesn't look like there's an index for this query, so one needs to be added in this PR.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/oauth2-client.js
Comment (dlongley) [reply lib/oauth2-client.js]: > (Our client management libraries (or anybody's, really), don't juggle different tokens (from the same issuer) based on scope.

Heh, the ambient authority and blast surface damage problems with oauth 2 in a nutshell. :P

File: lib/oauth2-client.js
Comment (mandyvenables) [reply lib/oauth2-client.js:59]: Thanks for the explanation, makes sense!

Added a comment about the key format.

File: lib/storage.js
Comment (mandyvenables) [reply lib/storage.js:41]: Added.

File: lib/authzHttpClient.js
Comment (mandyvenables) [reply lib/authzHttpClient.js]: This revealed a missing line of code, which is adding the newly generated access token to the authorization header. That code has been added.

File: lib/authzHttpClient.js
Comment (mandyvenables) [reply lib/authzHttpClient.js]: @dlongley How am I able to test this? In order to get this far I would need to have valid oauth2 credentials, and I don't think we have generic testing ones.

Comment (dlongley) [reply lib/authzHttpClient.js]: Great question! Probably some mock endpoints that expect certain mock tokens would make it easier to test. cc: @aljones15, @dmitrizagidulin -- suggestions?

Comment (mandyvenables) [reply lib/authzHttpClient.js]: Added 2 tests- one for success and one for failure with an incorrect access token.

File: lib/authzHttpClient.js
Comment (dmitrizagidulin) [inline lib/authzHttpClient.js]: Add `put` to the allowed proxy methods. (It's being added to `http-client` as well).

File: lib/authzHttpClient.js
Comment (dmitrizagidulin) [inline lib/authzHttpClient.js:41]: Add params docstring. Things like:

```
 * @param {object} options - Options hashmap.
 * @param {function} options.target - httpClient method ('get', 'post', etc).
 * @param {object} options.thisArg - httpClient instance.
 * @param {Array<*>} options.args - Method arguments ([url, options]).
```
etc

Comment (mandyvenables) [reply lib/authzHttpClient.js:41]: Added.

File: lib/authzHttpClient.js
Comment (mandyvenables) [reply lib/authzHttpClient.js]: Updated.

File: lib/main.js
Comment (dmitrizagidulin) [inline lib/main.js:19]: Do we still intend to process.exit() on non-retryable errors?

File: lib/oauth2-client.js
Comment (dmitrizagidulin) [inline lib/oauth2-client.js:112]: I think we need more details here than 'Service Unavailable', for our own debugging purposes.
At very least, we should say something like "Could not get access token from service `${token_endpoint}` after X retries."

Comment (dmitrizagidulin) [reply lib/oauth2-client.js:112]: Also, let's double-check that this bedrock error is logged to the `logger` we use elsewhere in the code.

File: lib/main.js
Comment (mandyvenables) [reply lib/main.js:19]: Yes, the exit occurs here: https://github.com/digitalbazaar/bedrock-oauth2-client/blob/main/lib/oauth2-client.js#L184

File: lib/oauth2-client.js
Comment (mandyvenables) [reply lib/oauth2-client.js:112]: @dmitrizagidulin There are multiple errors that are sent to `logger` throughout the process, including the initial error when trying to get an access token, and any subsequent errors on retries, and finally what the unrecoverable error was (if there was one). What other information should be logged?

File: lib/oauth2-client.js
Comment (dlongley) [inline lib/oauth2-client.js]: I think we usually go with this style:

```suggestion
    throw new TypeError(`"client_id" must be a string.`);
```

File: lib/oauth2-client.js
Comment (dlongley) [inline lib/oauth2-client.js]:
```suggestion
    throw new TypeError(`"client_secret" must be a string.`);
```

File: lib/oauth2-client.js
Comment (dlongley) [inline lib/oauth2-client.js]:
```suggestion
    throw new TypeError(`"token_endpoint" must be a string.`);
```

File: lib/oauth2-client.js
Comment (dlongley) [inline lib/oauth2-client.js]: Do these need to happen in serial or can they be in parallel?

File: lib/storage.js
Comment (dlongley) [inline lib/storage.js]:
```suggestion
    // index for query that is used to look up tokens
```

File: lib/storage.js
Comment (dlongley) [inline lib/storage.js]: Now that we have this index, do we need the `token.token` unique index? It seems like it could even technically be "ok" if two tokens with the same value were stored if they were for a different client + endpoint (even though if this happened it would likely be an indication of some problem with the authz servers).

Comment (dlongley) [reply lib/storage.js]: cc: @dmitrizagidulin

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/storage.js
Comment (dmitrizagidulin) [reply lib/storage.js]: Agreed, we do not need that index to be unique.

Comment (mandyvenables) [reply lib/storage.js]: Updated.

File: lib/oauth2-client.js
Comment (dmitrizagidulin) [reply lib/oauth2-client.js:112]: Ok, excellent (re logger), was just checking to make sure.

I do think the error itself should have more information, though (since that's what's going to be returned to the calling app). So, for example:

Our call chain is --

Client caller `-->` Service 1  `-->` Service 2 (tries to renew access token from) `-->` OAuth2 IDP.

So, we know that we don't want a lot of detail to return to the client.
But it does make sense (for traceability) to be able to tell from the Service 1 error log, that the 503 Service Unavailable error it received from Service 2 was caused by it not being able to get a token after 3 retries.

@mattcollier, @dlongley - what do you think?

Comment (mattcollier) [reply lib/oauth2-client.js:112]: +1 to minimal return to the caller client and +1 to logging usable error that does not reveal any secrets into the the application logs.

File: lib/oauth2-client.js
Comment (dlongley) [inline lib/oauth2-client.js]: Any reason not to simplify to this?

```suggestion
    promises.push(_createAccessToken(clientInfo));
```

File: lib/storage.js
Comment (dlongley) [inline lib/storage.js]: I'd move this index up first so it's the first one people read -- it's the most important one for understanding the data modeling. So this one first, then the expires one. We generally try to order indexes that way (or if there are many, perhaps alphabetically -- but either way, this second one should be first).

**Review by dlongley — APPROVED** [review]
Thanks! Ok, AFAICT, this is now achieving what we set out to do now. There's a minor comment or two from me to address and some comments from others -- so we'll want to get their approval too before merging.

File: lib/oauth2-client.js
Comment (mandyvenables) [reply lib/oauth2-client.js]: Updated.

File: lib/oauth2-client.js
Comment (mandyvenables) [reply lib/oauth2-client.js:112]: @dmitrizagidulin I added "name" (which comes from the `config.authorization`) and specifies the service, so the error in the logger would be `logger.debug(`Failed to acquire access token for "${name}".`, {e});`

Is that enough, or should the 503 error be changed as well?

Comment (dmitrizagidulin) [reply lib/oauth2-client.js:112]: Yeah, lets add details to the 503 error as well.

File: lib/oauth2-client.js
Comment (dmitrizagidulin) [inline lib/oauth2-client.js]: I think this should be more like `Service Unavailable: Could not renew token for ${audience}`

**Review by dmitrizagidulin — APPROVED** [review]
:+1: 
@mandyvenables thanks!

**Pull request merged**
